torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -137,15 +137,29 @@ std::unique_ptr<RpcCommandBase> RequestCallbackImpl::processRpc(
       // Attach 'recv' autograd function.
       DistAutogradContext* autogradContext = addRecvRpcBackward(
           rpcWithAutograd.autogradMetadata(), rpcWithAutograd.tensors());
+      // For this recv thread, before processRpc(), set current_context_id_ to
+      // be context_id passed from client. In this way, if there is nested
+      // rpc call in python rpc call, original context_id from client can be
+      // passed in the chain calls.
+      auto& autogradContainer = DistAutogradContainer::getInstance();
+      if (autogradContext != nullptr) {
+        autogradContainer.set_current_context_id(autogradContext->context_id());
+      }
 
       // Process the original RPC.
       auto wrappedMessageType = rpcWithAutograd.wrappedMessageType();
       auto wrappedRpcResponse =
           processRpc(rpcWithAutograd.wrappedRpc(), wrappedMessageType);
 
+      // After processRpc() is done, clean up current_context_id_ to be -1,
+      // for a recv thread, current_context_id_ should always be invalid after
+      // processRpc() is done.
+      if (autogradContext != nullptr) {
+        autogradContainer.set_current_context_id(-1);
+      }
```

Comment:
We should clear the context id in `request_callback.cpp` in the operator() function (since we're guaranteed every RPC would go through here). We can do something like this:

```
Message RequestCallback::operator()(Message& request) const {
  ClearAutogradContextGuard guard;
  try {
    return processMessage(request);
  } catch (std::exception& e) {
    LOG(ERROR) << "Received error while processing request type "
               << request.type() << ": " << e.what();
    return createException(request, e);
  }
}
```

Basically in the destructor of `ClearAutogradContextGuard` we set the current context id to -1. Also, we shouldn't pass -1 here like this. The value -1 is an internal detail of the `DistAutogradContext` class. We should add a method called `clearCurrentContext()` which sets the value to -1.